filebeat/input/tcp/client.go

```diff
@@ -0,0 +1,118 @@
+package tcp
+
+import (
+	"bufio"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Client is a remote client.
+type Client struct {
+	conn          net.Conn
+	forwarder     *harvester.Forwarder
+	done          chan struct{}
+	metadata      common.MapStr
+	splitFunc     bufio.SplitFunc
+	maxReadBuffer int64
+	timeout       time.Duration
+}
+
+// NewClient returns a new client instance for the remote connection.
+func NewClient(
+	conn net.Conn,
+	forwarder *harvester.Forwarder,
+	splitFunc bufio.SplitFunc,
+	maxReadBuffer int64,
+	timeout time.Duration,
+) *Client {
+	client := &Client{
+		conn:          conn,
+		forwarder:     forwarder,
+		done:          make(chan struct{}),
+		splitFunc:     splitFunc,
+		maxReadBuffer: maxReadBuffer,
+		timeout:       timeout,
+	}
+	client.cacheMetadata()
+	return client
+}
+
+// Handle is reading message from the specified TCP socket.
+func (c *Client) Handle() error {
+	defer c.conn.Close()
+
+	r := NewMeteredReader(NewDeadlineReader(c.conn, c.timeout), c.maxReadBuffer)
+	buf := bufio.NewReader(r)
+	scanner := bufio.NewScanner(buf)
+	scanner.Split(c.splitFunc)
+
+	for scanner.Scan() {
+		if scanner.Err() != nil {
+			// we are forcing a close on the socket, lets ignore any error that could happen.
+			select {
+			case <-c.done:
+				break
+			default:
+			}
+			// This is a user defined limit and we should notify the user.
+			if IsMaxReadBufferErr(scanner.Err()) {
+				logp.Err("tcp client error: %s", scanner.Err())
+			}
+			return errors.Wrap(scanner.Err(), "tcp client error")
+		}
+		r.Reset()
+		// TODO(ph), should we keep the byte array instead of allocating a string?
+		c.forwarder.Send(c.createEvent(scanner.Text()))
```

Comment:
This is a legitimate question, I am not sure how the remaining part of the code will work if I keep the message as a byte array.

Reply:
I would ignore the overhead for now if it adds code complexity.

Reply:
For me on my side there is no overhead, it just a matter of calling `scanner.Bytes()` instead of `scanner.Text()`, I am just unaware of the downstream consequence for processor If I keep the bytes.

Reply:
Actually, I just read the [doc](https://golang.org/src/bufio/scan.go), It's better to do an allocation because the original bytes can be overridden.

```golang
// Bytes returns the most recent token generated by a call to Scan.
  // The underlying array may point to data that will be overwritten
  // by a subsequent call to Scan. It does no allocation.
  func (s *Scanner) Bytes() []byte {
  	return s.token
  }
```